db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OperatorNode.java

```diff
@@ -32,16 +32,319 @@
+                    op.treePrint(depth + 1);
+                }
+            }
+        }
+    }
+
+    public void bindOperands(FromList fromList,
+                             SubqueryList subqueryList,
+                             List<AggregateNode>    aggregateVector) throws StandardException
+    {
+        for (int i = 0; i < operands.size(); ++i) {
+            if (operands.get(i) != null) {
+                operands.set(i, operands.get(i).bindExpression(fromList, subqueryList, aggregateVector));
+            }
+        }
+    }
+
+
+    /** generate a SQL->Java->SQL conversion tree above the
+     * operands of this Generic Operator Node if needed. Subclasses can override
+     * the default behavior.
+     */
+    public ValueNode genSQLJavaSQLTree() throws StandardException
+    {
+        for (int i = 0; i < operands.size(); ++i) {
+            TypeId typeId = operands.get(i).getTypeId();
+            if (typeId.userType())
+                operands.set(i, operands.get(i).genSQLJavaSQLTree());
+        }
+
+        return this;
+    }
+
+    /**
+     * Preprocess an expression tree.  We do a number of transformations
+     * here (including subqueries, IN lists, LIKE and BETWEEN) plus
+     * subquery flattening.
+     * NOTE: This is done before the outer ResultSetNode is preprocessed.
+     *
+     * @param    numTables            Number of tables in the DML Statement
+     * @param    outerFromList        FromList from outer query block
+     * @param    outerSubqueryList    SubqueryList from outer query block
+     * @param    outerPredicateList    PredicateList from outer query block
+     *
+     * @return        The modified expression
+     *
+     * @exception StandardException        Thrown on error
+     */
+    public ValueNode preprocess(int numTables,
+                                FromList outerFromList,
+                                SubqueryList outerSubqueryList,
+                                PredicateList outerPredicateList)
+            throws StandardException
+    {
+        for (int i = 0; i < operands.size(); ++i) {
+            if (operands.get(i) != null) {
+                operands.set(i, operands.get(i).preprocess(
+                        numTables,
+                        outerFromList, outerSubqueryList,
+                        outerPredicateList));
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public boolean checkCRLevel(int level){
+        return operands.stream().anyMatch(op -> op != null && op.checkCRLevel(level));
+    }
+
+
+    /**
+     * Remap all ColumnReferences in this tree to be clones of the
+     * underlying expression.
+     *
+     * @return ValueNode            The remapped expression tree.
+     *
+     * @exception StandardException            Thrown on error
+     */
+    public ValueNode remapColumnReferencesToExpressions()
+            throws StandardException
+    {
+        for (int i = 0; i < operands.size(); ++i) {
```

Comment:
does it make sense to check that `operands` is not null? (same in other methods such as `checkCRLevel` and `isConstantExpression` ... etc).

Reply:
Done